File: 215.kth-largest-element-in-an-array.py
#
# @lc app=leetcode id=215 lang=python
#
# [215] Kth Largest Element in an Array
#

# @lc code=start
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        return self.quickSelect(nums, 0, len(nums)-1, k)

    def quickSelect(self, nums, start, n, k):
        pos = self.partition(nums, start, n)
        if pos == k-1:
            return nums[pos]
        elif pos >= k:
            return self.quickSelect(nums, start, pos-1, k)
        else:
            return self.quickSelect(nums, pos+1, n, k)

    def partition(self, nums, left, right):
        pivot = nums[right]
        i = left
        for j in range(left, right):
            if nums[j] > pivot:
                nums[j], nums[i] = nums[i], nums[j]
                i += 1
        nums[right], nums[i] = nums[i], nums[right]
        return i
        
# @lc code=end

File: 215.kth-largest-element-in-an-array.cpp
/*
 * @lc app=leetcode id=215 lang=cpp
 *
 * [215] Kth Largest Element in an Array
 */
#include <iostream>
#include <vector>
using namespace std;

// @lc code=start
class Solution
{
public:
    int findKthLargest(vector<int> &nums, int k)
    {
        k = nums.size() - k;
        int l = 0, h = nums.size() - 1;
        while (l < h)
        {
            int j = partition1(nums, l, h);
            if (j == k)
                break;
            else if (j < k)
                l = j + 1;
            else
                h = j - 1;
        }
        return nums[k];
    }

    int partition(vector<int> &nums, int l, int h)
    {
        int pivot = nums[l];
        int i = l, j = h + 1;
        while (i < j)
        {
            while (i < h && nums[++i] < pivot)
                ;
            while (j > l && nums[--j] > pivot)
                ;
            if (i >= j)
                break;
            swap(nums[i], nums[j]);
        }
        swap(nums[j], nums[l]);
        return j;
    }

    int partition1(vector<int> &nums, int l, int h)
    {
        int pivot = nums[h];
        int i = l - 1;
        for (int j = l; j < h; j++)
        {
            if (nums[j] < pivot)
            {
                swap(nums[++i], nums[j]);
            }
        }
        swap(nums[++i], nums[h]);
        return i;
    }
};
// @lc code=end
int main()
{
    vector<int> a = {3, 2, 1, 5, 6, 4};
    int r = Solution().findKthLargest(a, 2);
    cout << r << endl;
}
